### PR TITLE
feature: add watch function

### PR DESCRIPTION
**Pull request by Jeremy-Run**
att

### CODE REVIEW

File: batch.go [hunk 1]
```diff
@@ -272,8 +272,18 @@ func (b *Batch) Commit() error {
 	for key, record := range b.pendingWrites {
 		if record.Type == LogRecordDeleted {
 			b.db.index.Delete(record.Key)
+			// insert delete event
```
Comment (roseduan) [inline batch.go:275]: `NewEvent` function seems redundant, just insert it directly by
```
b.db.watcher.Insert(&Event{...})
```

File: db.go [hunk 2]
```diff
@@ -45,6 +45,8 @@ type DB struct {
 	closed       bool
 	mergeRunning uint32 // indicate if the database is merging
 	batchPool    sync.Pool
+	watch        chan *Event // user consume channel for watch events
```
Comment (roseduan) [inline db.go:48]: `watchChan` or `watchCh` may be more clear

File: options.go [hunk 3]
```diff
@@ -27,6 +27,11 @@ type Options struct {
 
 	// BytesPerSync specifies the number of bytes to write before calling fsync.
 	BytesPerSync uint32
+
+	// Watchable whether to enable event change notification.
+	Watchable bool
+	// WatchQueueSize the cache length of the watch queue, default 1000.
```
Comment (roseduan) [inline options.go:33]: Can we just use one option WatchQueueSize, if the size is greater than 0, which means enable the watch?

File: watch.go [hunk 4]
```diff
@@ -0,0 +1,112 @@
+package rosedb
+
+import (
+	"sync"
+	"time"
+)
+
+const (
+	ActionPut    = "Put"
```
Comment (roseduan) [inline watch.go:9]: Can we use a alias for this?
```
type WatchActionType = byte
const (
    WatchActionPut WatchActionType = iota
    WatchActionDelete
)
```

File: watch.go [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+// Event
+type Event struct {
+	Action  string
+	Key     []byte
+	Value   []byte
+	BatchId uint64
+}
+
+// NewEvent generate a key-value change event after batch submission.
+func NewEvent(action string, key, value []byte, batchId uint64) *Event {
```
Comment (roseduan) [inline watch.go:22]: Not ncessary, refer to the first comment

File: watch.go [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+	}
+	return &Watcher{
+		queue: eventQueue{
+			Events:   make([]*Event, capacity),
+			Capacity: capacity,
+		},
+	}
+}
+
+func (w *Watcher) Insert(e *Event) {
```
Comment (roseduan) [inline watch.go:53]: This function will not exposed to caller? if so, the first letter will be lowercase.
`insertEvent` may be better.

File: watch.go [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+	w.mu.Lock()
+	w.queue.push(e)
+	if w.queue.isFull() {
+		w.queue.frontTakeAStep()
+	}
+	w.mu.Unlock()
+}
+
+func (w *Watcher) Scan() (e *Event, isEmpty bool) {
```
Comment (roseduan) [inline watch.go:62]: the name Scan is inappropriate, it is not aa iterator something. How about `GetEvent`?

and we if the returned event is null, which means empty queue, so the second return value is not necessary?

File: watch.go [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+	defer w.mu.RUnlock()
+	if isEmpty = w.queue.isEmpty(); isEmpty {
+		return
+	}
+	e = w.queue.pop()
+	return
+}
+
+// Sync synchronize events to DB's watch
+func (w *Watcher) Sync(c chan *Event) {
```
Comment (roseduan) [inline watch.go:73]: Sync -> sendEvent, should also be lowercase if not belongs to the caller.

**Comment by roseduan** [comment]
Thanks, please also add an example folder to show the caller how to use this new feature

**Comment by Jeremy-Run** [comment]
> Thanks, please also add an example folder to show the caller how to use this new feature

ok

**Comment by roseduan** [comment]
I will add more comments aftet I merge it, anyway, it looks very good, thanks!

**Comment by Jeremy-Run** [comment]
> I will add more comments aftet I merge it, anyway, it looks very good, thanks!

ok, you are welcome

**Pull request merged**